fix: :bug: updated git fetch to better detect the ancestry of the branches and correctly determine the validity of the branch merge request

BrickPLuM opened the pull request:
Update the way Git retrieves information (specifically ancestry) for a branch.

Fix for the error:

```
Run bricklanetech/action.control-merge@v1
/usr/bin/docker run --name f7c3612010521d0bc04c488155a96c1d946141_f1d795 --label f7c361 --workdir /github/workspace --rm -e "GITHUB_TOKEN" -e "INPUT_WORKFLOW" -e "INPUT_FEATURE_PATTERN" -e "INPUT_HOTFIX_PATTERN" -e "HOME" -e "GITHUB_JOB" -e "GITHUB_REF" -e "GITHUB_SHA" -e "GITHUB_REPOSITORY" -e "GITHUB_REPOSITORY_OWNER" -e "GITHUB_REPOSITORY_OWNER_ID" -e "GITHUB_RUN_ID" -e "GITHUB_RUN_NUMBER" -e "GITHUB_RETENTION_DAYS" -e "GITHUB_RUN_ATTEMPT" -e "GITHUB_ACTOR_ID" -e "GITHUB_ACTOR" -e "GITHUB_WORKFLOW" -e "GITHUB_HEAD_REF" -e "GITHUB_BASE_REF" -e "GITHUB_EVENT_NAME" -e "GITHUB_SERVER_URL" -e "GITHUB_API_URL" -e "GITHUB_GRAPHQL_URL" -e "GITHUB_REF_NAME" -e "GITHUB_REF_PROTECTED" -e "GITHUB_REF_TYPE" -e "GITHUB_WORKFLOW_REF" -e "GITHUB_WORKFLOW_SHA" -e "GITHUB_REPOSITORY_ID" -e "GITHUB_TRIGGERING_ACTOR" -e "GITHUB_WORKSPACE" -e "GITHUB_ACTION" -e "GITHUB_EVENT_PATH" -e "GITHUB_ACTION_REPOSITORY" -e "GITHUB_ACTION_REF" -e "GITHUB_PATH" -e "GITHUB_ENV" -e "GITHUB_STEP_SUMMARY" -e "GITHUB_STATE" -e "GITH
-> checking if hotfix
-> checking if merge is allowed in the workflow rules
--> source is feature branch and target either another feature or start of workflow
-> checking if branch is blocked
fatal: ambiguous argument 'origin/testing': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: ambiguous argument 'origin/production': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: Not a valid object name origin/testing
--> Branch is blocked
fatal: ambiguous argument 'origin/testing': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: ambiguous argument 'origin/testing': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
✘ Branch testing is awaiting merge into production, please check with 
'git <command> [<revision>...] -- [<file>...]'
```

BrickPLuM committed:
fix: :bug: updated git fetch to better detect the ancestry of the branches and correctly determine the validity of the branch merge request

davidjdixon commented on entrypoint.sh:
```diff
@@ -111,6 +111,14 @@ echo "FEATURE_PATTERN=$FEATURE_PATTERN"
 # mark repo directory as safe to prevent 'dubious ownership' detected in the repository
 git config --global --add safe.directory /github/workspace
 
+# change to workspace directory
+cd /github/workspace || exit 1
+
+# Fetch all remote branches with full history needed for merge-base comparison
+# GitHub Actions provides a shallow clone, so we need to unshallow and fetch all refs
+git fetch --unshallow origin '+refs/heads/*:refs/remotes/origin/*' 2>/dev/null || \
+    git fetch origin '+refs/heads/*:refs/remotes/origin/*' 2>/dev/null || true
+
```
@BrickPLuM could you follow the suggestion above as we should be reporting back issues rather than silencing them

BrickPLuM committed:
feat: :white_check_mark: add testing script and fixed error handling in the entrypoint script

BrickPLuM committed:
test: :white_check_mark: fix issues with test script not being flexible with different workflows

Can now detect workflow from clonex repository and accept an override where needed

BrickPLuM committed:
test: :white_check_mark: add more resilliance to workflow detections

davidjdixon reviewed (approved):
nice addition of the testing script! And this change should prevent the need for per-project `fetch-depth:0` additions which would make this more compatible with the old checkout workflow.
